### PR TITLE
3.1: Improve AD integration tests

### PR DESCRIPTION
**Pull request by gmarciani**
### Description of changes
1. Add testing for MicrosoftAD.
2. Add testing for LDAP.
3. Add testing for LDAPS with and without certificate verification.
4. Add testing for LdapAccessFilter
5. Execute user password reset within a retry with backoff block. With this change we are now more robust against failures in reset. Consider that a failed password reset would make the user disabled, thus causing apparently random test failures.
6. Fix ACM certificated deletion: it should be deleted only if no-delete option is missing
7. Add log line to explain 10min sleep time waiting for certificate setup in the head node.
8. Allow reusing of an existent IAM user role stack.
9. Fix typo in remote command executor log.

### Tests
1 Integration tests manually launched locally, both for SimpleAd and MsAD.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: tests/integration-tests/conftest.py [hunk 1]
```diff
@@ -805,14 +809,23 @@ def initialize_cli_creds(cfn_stacks_factory, request):
...
-        logging.info("Creating IAM roles for pcluster CLI")
-        stack_name = generate_stack_name("integ-tests-iam-user-role", request.config.getoption("stackname_suffix"))
-        stack_template_path = os.path.join("..", "iam_policies", "user-role.cfn.yaml")
-        with open(stack_template_path, encoding="utf-8") as stack_template_file:
-            stack_template_data = stack_template_file.read()
-        stack = CfnStack(name=stack_name, region=region, capabilities=["CAPABILITY_IAM"], template=stack_template_data)
-        cfn_stacks_factory.create_stack(stack)
+        if request.config.getoption("iam_user_role_stack_name"):
+            stack_name = request.config.getoption("iam_user_role_stack_name")
+            logging.info("Using stack {0} in region {1}".format(stack_name, region))
```
Comment (enrico-usai) [inline tests/integration-tests/conftest.py]: minor: you can use lazy evaluation on logging strings

File: tests/integration-tests/tests/ad_integration/test_ad_integration.py [hunk 2]
```diff
@@ -595,3 +647,20 @@ def test_ad_integration(
     for user in users:
         user.reset_stateful_connection_objects(remote_command_executor)
     _check_ssh_key_generation(users[1], scheduler_commands, True)
+    for user in users:
+        logging.info(f"Checking SSH access for user {user.alias}")
+        if user.alias == "PclusterUser3":
+            _check_ssh_auth(user=user, expect_success=False)
+        else:
+            _check_ssh_auth(user=user, expect_success=True)
```
Comment (enrico-usai) [inline tests/integration-tests/tests/ad_integration/test_ad_integration.py]: minor, you can do:
```python
expect_success = user.alias != "PclusterUser3"
_check_ssh_auth(user=user, expect_success=False)
```

or
```python
_check_ssh_auth(user=user, expect_success=user.alias != "PclusterUser3")
```

**Review by enrico-usai — DISMISSED** [review]

**Review by enrico-usai — DISMISSED** [review]

File: tests/integration-tests/tests/ad_integration/test_ad_integration.py [hunk 3]
```diff
@@ -58,19 +58,44 @@ def get_infra_stack_parameters(stack_name):
...
     ldap_search_base = ",".join(
         [f"dc={domain_component}" for domain_component in directory_stack_outputs.get("DomainName").split(".")]
     )
+    domain_short_name = directory_stack_outputs.get("DomainShortName")
     read_only_username = directory_stack_outputs.get("ReadOnlyUserName")
+
+    if directory_type == "SimpleAD":
+        ldap_default_bind_dn = f"CN={read_only_username},CN=Users,{ldap_search_base}"
+    elif directory_type == "MicrosoftAD":
+        ldap_default_bind_dn = f"CN={read_only_username},OU=Users,OU={domain_short_name},{ldap_search_base}"
```
Comment (hanwen-pcluste) [inline tests/integration-tests/tests/ad_integration/test_ad_integration.py:80]: So SimpleAD uses tow `CN=` while Microsoft AD uses `CN=` and `OU=`?

**Review by hanwen-pcluste — APPROVED** [review]
Awesome!

**Pull request merged**